[feature] add expressions at the symbollist level

vmora opened the pull request:
Size and Rotation can be defined by an expression for all symbols
composing a marker Width can be defined by an expression for all symbols
composing a line

For marquers, a legend is generated for varying sizes. This allows
multivariate analysis legend in the case of classified/graduated colors.

The line symbol layer as been refactored to avoid code duplication and
expose the offset and offset units in the base class. This refacto was
necessary for this feature to avoid unnecessary dynamic cast. Note that
the added functions in the base class where already defined in all child
classes.

The offset is now set along with size to maintain the relative position
of symbols composing a marker.

An asistant, with preview, is accessible through the data defined button
to help the user define the size expression. Three methods are
available: Frannery, Area and Radius.

Added a widged for use in categorized/classified symbology gui to set
the expression if needed. The assistant is also available from it.

Note: breaking change in the symbolslist widget. The widget ctor now
takes the layer as a parameter, this is necessary for the assistant and
for the data-defined buttons. The widget is used only in one place and
although it's a change to the public interface, the likelyhood of
breaking plugins dependent on this widget seems low enought.

vmora committed:
[feature] add expressions at the symbollist level

Size and Rotation can be defined by an expression for all symbols
composing a marker Width can be defined by an expression for all symbols
composing a line

For marquers, a legend is generated for varying sizes. This allows
multivariate analysis legend in the case of classified/graduated colors.

The line symbol layer as been refactored to avoid code duplication and
expose the offset and offset units in the base class. This refacto was
necessary for this feature to avoid unnecessary dynamic cast. Note that
the added functions in the base class where already defined in all child
classes.

The offset is now set along with size to maintain the relative position
of symbols composing a marker.

An asistant, with preview, is accessible through the data defined button
to help the user define the size expression. Three methods are
available: Frannery, Area and Radius.

Added a widged for use in categorized/classified symbology gui to set
the expression if needed. The assistant is also available from it.

Note: breaking change in the symbolslist widget. The widget ctor now
takes the layer as a parameter, this is necessary for the assistant and
for the data-defined buttons. The widget is used only in one place and
although it's a change to the public interface, the likelyhood of
breaking plugins dependent on this widget seems low enought.

vmora commented:
replaces #1961

vmora commented:
fixing...

nyalldawson commented on src/core/layertree/qgslayertreemodellegendnode.cpp:
```diff
@@ -187,11 +188,67 @@ QVariant QgsSymbolV2LegendNode::data( int role ) const
+            {
+              if ( qAlpha( img.pixel( x, y ) ) )
+              {
+                xmin = x < xmin ? x : xmin;
```
May as well use qMin/qMax here

vmora replied:
done

nyalldawson commented on src/core/layertree/qgslayertreemodellegendnode.cpp:
```diff
@@ -187,11 +188,67 @@ QVariant QgsSymbolV2LegendNode::data( int role ) const
+        // crop
+        if ( mItem.symbol()->type() == QgsSymbolV2::Marker && mCrop )
+        {
+          pix = QgsSymbolLayerV2Utils::symbolPreviewPixmap( mItem.symbol(), QSize( 512, 512 ), validData ? &context : 0 );
```
this section should be moved out to its own static function in QgsSymbolLayerV2Utils, it's a useful routine to have widely available

vmora replied:
done, function crop in QgsSymbolLayerV2Utils

nyalldawson commented on src/core/layertree/qgslayertreemodellegendnode.cpp:
```diff
@@ -187,11 +188,67 @@ QVariant QgsSymbolV2LegendNode::data( int role ) const
+          QImage img = pix.toImage();
+          // find the non-transparent area bbox
+          int ymin = img.height(), ymax = 0;
+          for ( int y = 0; y < img.height(); y++ )
```
this loop makes the assumption that a line symbol is constant along it's length. That's not necessarily the case, eg for a marker line symbol.

vmora replied:
fixed with a loop on x too.

vmora replied:
now uses function crop, wich is the same as the fix and avoids duplication.

nyalldawson commented on src/core/layertree/qgslayertreemodellegendnode.h:
```diff
@@ -162,6 +162,11 @@ class CORE_EXPORT QgsSymbolV2LegendNode : public QgsLayerTreeModelLegendNode
 
     virtual void invalidateMapBasedData() override;
 
+    void setIconSize( const QSize& sz ) { mIconSize = sz; }
```
these methods need documentation

vmora replied:
done

nyalldawson commented on src/core/layertree/qgslayertreemodellegendnode.h:
```diff
@@ -162,6 +162,11 @@ class CORE_EXPORT QgsSymbolV2LegendNode : public QgsLayerTreeModelLegendNode
+    void setIconSize( const QSize& sz ) { mIconSize = sz; }
+    QSize iconSize() const { return mIconSize; }
+
+    void setCrop( bool isCropRequired ) { mCrop = isCropRequired; }
```
I'd suggest adding a getter for mCrop too. There's nothing more frustrating then a missing getter when your project depends on it ;)

vmora replied:
done

nyalldawson commented on src/core/qgsdatadefined.cpp:
```diff
@@ -43,6 +43,18 @@ QgsDataDefined::QgsDataDefined( const QgsExpression * expression )
+{
+  QgsExpression expression( expressionString );
+  mActive = expression.rootNode();
+  mUseExpression = expression.rootNode() && !dynamic_cast<const QgsExpression::NodeColumnRef*>( expression.rootNode() );
```
I'd suggest moving this check to something like QgsExpression::isSingleField() , and adding a few unit tests for it.

nyalldawson commented on src/core/qgsscaleexpression.cpp:
```diff
@@ -0,0 +1,103 @@
+  switch ( mType )
+  {
+    case Linear: return mMinSize + ( qBound( mMinValue, value, mMaxValue ) - mMinValue ) * ( mMaxSize - mMinSize ) / ( mMaxValue - mMinValue );
+    case Area: return pow( qBound( mMinValue, value, mMaxValue ) - mMinValue, .5 ) * ( mMaxSize - mMinSize ) / pow( mMaxValue - mMinValue, .5 );
```
should be qPow instead of pow

vmora replied:
done

nyalldawson commented on src/core/symbology-ng/qgssymbollayerv2utils.h:
```diff
@@ -426,6 +426,13 @@ class CORE_EXPORT QgsSymbolLayerV2Utils
+     *  The values are chosen so that they are 1, 2 or 5 times a power of 10.
+     * @note added in 2.10
+     */
+    static QList<double> prettyBreaks( double minimum, double maximum, int classes );
```
needs to be added to sip bindings too

vmora replied:
done

nyalldawson commented:
@vmora Just a note, I'm only part way through this. It's a big PR :)

nyalldawson commented on src/gui/qgsdatadefinedbutton.h:
```diff
@@ -167,6 +181,12 @@ class GUI_EXPORT QgsDataDefinedButton: public QToolButton
+     * Set an assistant to define expression
+     * takes ownership
+     */
+    void setAssistant( QgsDataDefinedAssistant * assistant ) { mAssistant.reset( assistant ); }
```
probably the parent of mAssistant should be set to "this" here too. That will help ensure that the dialog gets displayed in the right position (eg for multi monitor setups)

vmora replied:
done

nyalldawson commented on src/gui/qgsdatadefinedbutton.h:
```diff
@@ -15,16 +15,30 @@
 
-#include "qgsfield.h"
+#include <qgsfield.h>
+#include <qgsdatadefined.h>
```
Is this needed in the header?

vmora replied:
nope, you're right. done

nyalldawson commented on src/gui/qgsdatadefinedbutton.cpp:
```diff
@@ -394,6 +404,21 @@ void QgsDataDefinedButton::showDescriptionDialog()
+{
+  if ( mAssistant->exec() == QDialog::Accepted )
+  {
+    const QString newExp = mAssistant->expressionText();
```
Why not make QgsDataDefinedAssistant return a QgsDataDefined instead of just a string? That way assistants could be created which fields, deactivated expressions, etc. It would also be more future proof when QgsDataDefined is extended to support additional features (such as time based/animation support).

vmora replied:
done.

nyalldawson commented on src/gui/qgsdatadefinedbutton.h:
```diff
@@ -15,16 +15,30 @@
 class QgsDataDefined;
 
 /** \ingroup gui
+ * \class Qgs
```
All these added methods/classes need to be added to the SIP bindings too, and should have "added in ..." notes.

nyalldawson commented:
@vmora I think we could also speed up merging this if we commit it piece-by-piece. Eg, I **really** like the "assistant" changes you've made to QgsDataDefinedButton, so if the small fixes noted were made to that then we could merge that portion right away.

nyalldawson commented on src/core/symbology-ng/qgssymbollayerv2utils.cpp:
```diff
@@ -3781,3 +3781,146 @@ QString QgsSymbolLayerV2Utils::fieldOrExpressionFromExpression( QgsExpression* e
   return expression->expression();
 }
 
+QList<double> QgsSymbolLayerV2Utils::prettyBreaks( double minimum, double maximum, int classes )
```
this change (moving _prettyBreaks) to QgsSymbolLayerV2Utils could be split into it's own commit - that way we could push these changes straight away and help decrease the size of this PR

nyalldawson commented:
@vmora I've just committed some portions of this directly, to help reduce the size of this PR. There's a lot  happening here and I'm having trouble getting my head around some of it. Hope that's OK! Can you rebase this branch against master and I'll review the remaining changes?

Thanks!

vmora commented:
@nyalldawson Done.

vmora commented:
@nyalldawson oups, not yet

vmora commented:
@nyalldawson now done, really :)

nyalldawson commented:
@vmora I've just pushed some more components of this, thanks! I made a few changes to QgsScaleExpression:
- ran the code through scripts/prepare-commit.sh to fix spacing
- fixed the copyright header
- added python bindings
- also fixed some of the maths in QgsScaleExpression::size - some of the types weren't adding mMinSize to the returned value
- some other QString related optimisations

Hope that's ok! Don't worry about rebasing after this, there's not too much left to sort through

nyalldawson commented on src/core/symbology-ng/qgssymbolv2.h:
```diff
@@ -280,10 +281,26 @@ class CORE_EXPORT QgsMarkerSymbolV2 : public QgsSymbolV2
+
+    /** Set "en masse" expression for size
+     */
+    void setSizeExpression( const QString & exprStr );
```
these methods (setSizeExpression, sizeExpression, setWidthExpression, widthExpression) should accept/return a QgsDataDefined rather than QStrings. This will avoid the current situation where deactivating the size/width data defined button results in lose of the expression, and will future proof this work for when QgsDataDefined is enhanced in future (eg, there's plans for QgsDataDefined to allow for animation support, etc).

vmora replied:
Hi @nyalldawson, 

I don't mind changing the API, but the underlying storage is expressions, so it won't fix the problem right away. Or do you suggest I change mDataDefinedProperties to be a map of QgsDataDefined ? Because that's a huge change.

As a side note, I think that some parts of QgsDataDefined are necessary (like activate/deactivate) but some parts should be handled by expression, like deciding if it's a simple field or imposing the return type (improvement of qgsexpression required).

nyalldawson replied:
@vmora for new methods like the en masse size these should definitely be stored as QgsDataDefined. There's two main reasons for this - one is that the behaviour of QgsDataDefinedButtons should be consistent between symbology and labeling/composer, and this requires storing the extra information like active/deactive which requires QgsDataDefined. The more important reason though is that it future proofs this work. I'm currently working on enhancements to QgsDataDefined to allow a system of inheritable properties which are evaluated against expression contexts (you can see some early discussion here https://github.com/qgis/QGIS-Enhancement-Proposals/pull/12 and an early version of the expression context work here https://github.com/nyalldawson/QGIS/tree/exp_context8 and here https://github.com/nyalldawson/QGIS/commit/cb5f244fc92c60418a25f8fa45fae4b95fb717c8 ). This work will make implementing things like this PR much easier, but unfortunately it's still a while off. In the meantime all new methods should utilise QgsDataDefined so they'll be ready to flip across.

I'm more than happy to help with this -- if it's not done now it will cause me a lot of work in future ;)

(btw, I started work on porting symbology to use QgsDataDefined some time ago, but haven't had a chance to finish this - the work in progress is here https://github.com/nyalldawson/QGIS/commit/6126e09418b49ec8a2258a43a85849fb80148c0b ).

nyalldawson replied:
FYI - I'm working on fixing up that switch to QgsDataDefined for symbology. Should be committed sometime in the next 48 hours, so maybe wait for that to land as it should simplify this...

nyalldawson replied:
@vmora ok, I've pushed that work. Let me know if you need a hand porting this code over to QgsDataDefined

vmora replied:
@nyalldawson I fixed the missing default value and made the changes recommender by @wonder-sk today.

I'm going to rebase the code against master and use QgsDataDefined.

nyalldawson commented on src/core/symbology-ng/qgssymbolv2.cpp:
```diff
@@ -35,6 +37,33 @@
 #include <cmath>
 
+inline
+QString rotateEnMasse( double additionalRotation, const QString & exprStr )
```
as per an earlier version of this PR, can these "enMasse" methods/classes all be renamed to avoid the use of "enMasse"? Something like "rotateWholeSymbol" or "rotateAllSymbolLayers" would be easier to understand and result in a nicer API

vmora replied:
done

nyalldawson commented on src/gui/symbology-ng/qgssymbolslistwidget.h:
```diff
@@ -30,7 +30,7 @@ class GUI_EXPORT QgsSymbolsListWidget : public QWidget, private Ui::SymbolsListW
 
   public:
-    QgsSymbolsListWidget( QgsSymbolV2* symbol, QgsStyleV2* style, QMenu* menu, QWidget* parent );
+    QgsSymbolsListWidget( const QgsVectorLayer * layer, QgsSymbolV2* symbol, QgsStyleV2* style, QMenu* menu, QWidget* parent );
```
unfortunately this API break isn't allowed. You'll need to make the layer optional and add checks to qgssymbolslistwidget.cpp so that all references to mLayer are optional. Looks like this only affects the data defined buttons, so those could be either disable or hidden if no layer is specified...

vmora replied:
done

nyalldawson commented on src/gui/symbology-ng/qgssymbolv2selectordialog.cpp:
```diff
@@ -110,31 +110,31 @@ class SymbolLayerItem : public QStandardItem
     QVariant data( int role ) const override
     {
       if ( role == Qt::DisplayRole || role == Qt::EditRole )
+  {
```
remember to run scripts/prepare_commit.sh to fix the spacing before committing, that should avoid unnecessary changes like these

vmora replied:
I usually do, sorry for that. Fixed.

nyalldawson commented on src/core/layertree/qgslayertreemodellegendnode.h:
```diff
@@ -162,6 +162,18 @@ class CORE_EXPORT QgsSymbolV2LegendNode : public QgsLayerTreeModelLegendNode
 
     virtual void invalidateMapBasedData() override;
 
+    //! Set the icon size (icon may be smaller if croping is enabled)
```
@wonder-sk do you mind looking at these legend related changes? ( QgsMapLayerLegend + QgsLayerTreeModelLegendNode)

vmora committed:
[feature] add expressions at the symbollist level
fixes issues noted in PR:
- enMasse renamed to WholeSymbol
- vector layer made optional for QgsSymbolsListWidget
- fixed spacing
- restored qgssymbollayerv2utils.sip qgsdatadefined.sip qgsexpression.h
qgsdatadefinedbutton.h and qgsdatadefined.h
- removed crop from qgssymbollayerv2utils

wonder-sk commented on src/core/qgsmaplayerlegend.cpp:
```diff
@@ -198,11 +198,33 @@ QList<QgsLayerTreeModelLegendNode*> QgsDefaultVectorLayerLegend::createLayerTree
+  {
+    Q_ASSERT( widthMax > 0 );
+    n->setIconSize( QSize( widthMax + 2, n->iconSize().rheight() + 2 ) );
+    n->setCrop( false );
```
If I understand the code correctly, it temporarily turns on cropping to get the minimum size of each symbol, calculates the max width and finally sets the icon size for all icons to match the width. After initialization, cropping is always disabled.

- could you remove crop()/setCrop() from the API and just have a call minimumIconSize() to get that cropped icon size? I would prefer to keep such code away from data() implementation as it is unclear the cropping is enabled just temporarily.
- why +2 for icon sizes? (better always comment why magic numbers are used)
- how about rule-based renderer? it shows legend in a tree, so I guess there should be different max width for each sub-tree (otherwise one wide icon somewhere in the tree will make everything wide)

vmora replied:
@wonder-sk, you understand the code correctly, I removed rop()/setCrop() and replaced by minimumIconSize() and the code is a lot nicer like that (876603aa2).

I don't know how to handle the tree representation, could you give me some advice/code example please ?

nyalldawson commented on src/gui/symbology-ng/qgssymbolslistwidget.h:
```diff
@@ -30,7 +30,7 @@ class GUI_EXPORT QgsSymbolsListWidget : public QWidget, private Ui::SymbolsListW
 
   public:
-    QgsSymbolsListWidget( QgsSymbolV2* symbol, QgsStyleV2* style, QMenu* menu, QWidget* parent );
+    QgsSymbolsListWidget( QgsSymbolV2* symbol, QgsStyleV2* style, QMenu* menu, QWidget* parent, const QgsVectorLayer * layer );
```
layer will need a default value of 0 to avoid the API break

vmora replied:
Yes, off course, sorry for this mistake.

vmora committed:
[feature] add expressions at the symbollist level
improvement recommended in PR
- add default value for layer to qgssymbolslistwidget ctor
- removed crop from api of qgslayertreemodellegendnode, and replaced by
  minimumIconSize

wonder-sk commented:
@vmora for the legend tree representation... the legend nodes may have a parent legend node assigned (actually key (ID) of the parent rule) - that information is used to build the tree from the flat list of nodes. So one basically needs to group legend nodes with the same parent and do the calculation of icon size in those groups.

Hope that helps... the key is to use QgsLayerTreeModelLegendNode::data(ParentRuleKeyRole) to get the parents. So far only the rule-based renderer supports that.

vmora committed:
[feature] add expressions at the symbollist level
    improvement recommended in PR
max symbol width is now computed for common parents so that one wide
symbol in a tree representation won't increase all branches width

vmora commented:
@wonder-sk thanks, that helps a lot, 44ecead6abf takes care of that.

@nyalldawson the previous commit (changing the crop to minimumIconSize()) broke the size assistant. I'm fixing that.

vmora committed:
[feature] add expressions at the symbollist level
fix size assistant (broken by crop removal)

vmora commented:
@nyalldawson I fixed the assistant, but I don't see the legend for varying size anymore. I think it's related to the rebase against master with your change for QgsDatadefined.

vmora commented:
I found the problem: the mExpression creation is deferred now, so that the QgsSymbolLayerV2::dataDefinedProperty now return NULL even if there is a mExpressionString.

I'm changing the interface (and implementation to avoid the use of the deprecated symbollayer interface) and it'll fix the problem.

But I think that storing the string AND the expression in QgsDataDefined multiplies the states of of the class and this is not good IMHO. Same thing for the mField. I'd rather have only a QgsExpression (smart) pointer and a mActive flag, and use expression introspection (like QgsExpression::isSimpleField) instead of duplicating those functionalities in QgsDataDefined.

vmora commented:
PS: I understand that QgsExpression is not copyable (I'm the one who initially disabled the buggy cpy), but implementing the cpy is IMO a cleaner option.

nyalldawson commented:
@vmora it's on my TODO list (making QgsExpression & QgsDataDefined implicitly shared), but it's a long list ;)

vmora commented:
@nyalldawson I'm asking because I'm modifying the code for QgsSymbolV2::sizeExpression and the rest, and the QgsDataDefined interface is tricky because an expressionString can be defined while expression() will return null nonetheless.

vmora commented:
I'm trying to change the interface of setSizeExpression as you requested, but I'm stuck because I cannot return a pointer to a QgsDataDefined that does not exist in this case, and I certainly don't want to return an allocated value.

@nyalldawson can you tell me why you choose to pass QgsDataDefined by pointer with ownership transfer instead of passing by const & and copying.

nyalldawson commented:
manually merged in 1e46196937827d2a79f978b0ed67e97be9531c81 - thanks @vmora!

vmora commented:
The feature was funded by [Agence de l'Eau Adour Garonne](http://www.eau-adour-garonne.fr/fr/index.html).
